dns forwarding: T5687: Implement ECS settings for PowerDNS recursor

HollyGurza opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
Implement ECS settings for PowerDNS recursor

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://vyos.dev/T5687

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
dns forwarding
## Proposed changes
<!--- Describe your changes in detail -->
Add new params to `/run/powerdns/recursor.conf` and new CLI commands.
```
set service dns forwarding options ecs-add-for 
set service dns forwarding options ecs-ipv4-bits 
set service dns forwarding options edns-subnet-allow-list 
```
Add smoke test for this CLI command

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly

HollyGurza committed:
dns forwarding: T5687: Implement ECS settings for PowerDNS recursor

HollyGurza commented:
Should new options be described in the documentation?
https://docs.vyos.io/en/latest/configuration/service/dns.html

c-po commented:
> Should new options be described in the documentation? https://docs.vyos.io/en/latest/configuration/service/dns.html

Yes please

HollyGurza commented:
documentation pull request: https://github.com/vyos/vyos-documentation/pull/1256

dmbaturin commented on interface-definitions/service_dns_forwarding.xml.in:
```diff
@@ -735,6 +735,49 @@
+                <children>
+                  <leafNode name="ecs-add-for">
+                    <properties>
+                      <help>List of client netmasks for which EDNS Client Subnet will be added</help>
```
Maybe just "EDNS client subnet"?

Even though it's a multi-node, we generally don't say "list" or similar because for many such options it's common to have just one.

HollyGurza replied:
fixed

HollyGurza committed:
dns forwarding: T5687: Implement ECS settings for PowerDNS recursor

Fix option descriptions

c-po commented on interface-definitions/service_dns_forwarding.xml.in:
```diff
@@ -735,6 +735,49 @@
+                  <help>DNS server options</help>
+                </properties>
+                <children>
+                  <leafNode name="ecs-add-for">
```
Constraints missing, will add a commit to fix this in this PR

c-po committed:
dns forwarding: T5687: add missing constraints on ecs-add-for CLI node

Completion help suggests only IPv4 and IPv6 prefixes are supported, thus add a
proper constraint enforcing this.

c-po committed:
smoketest: T5687: simplify "dns forwarding" test setup

Commit eb76729d6324 ("dns forwarding: T5687: Implement ECS settings for PowerDNS
recursor") added a helper "_set_required_options()" method to reduce duplicate
code when setting up the base interface test.

This refactors the test class to call this code always in setUp() so we have
it written only once.

c-po commented:
```
cpo@LR1.wue3:~$ /usr/libexec/vyos/tests/smoke/cli/test_service_dns_forwarding.py
test_basic_forwarding (__main__.TestServicePowerDNS.test_basic_forwarding) ... ok
test_dns64 (__main__.TestServicePowerDNS.test_dns64) ... ok
test_dnssec (__main__.TestServicePowerDNS.test_dnssec) ... ok
test_domain_forwarding (__main__.TestServicePowerDNS.test_domain_forwarding) ... ok
test_ecs_add_for (__main__.TestServicePowerDNS.test_ecs_add_for) ... ok
test_ecs_ipv4_bits (__main__.TestServicePowerDNS.test_ecs_ipv4_bits) ... ok
test_edns_subnet_allow_list (__main__.TestServicePowerDNS.test_edns_subnet_allow_list) ... ok
test_exclude_throttle_adress (__main__.TestServicePowerDNS.test_exclude_throttle_adress) ... ok
test_external_nameserver (__main__.TestServicePowerDNS.test_external_nameserver) ... ok
test_listening_port (__main__.TestServicePowerDNS.test_listening_port) ... ok
test_no_rfc1918_forwarding (__main__.TestServicePowerDNS.test_no_rfc1918_forwarding) ... ok
test_serve_stale_extension (__main__.TestServicePowerDNS.test_serve_stale_extension) ... ok

----------------------------------------------------------------------
Ran 12 tests in 52.853s

OK
```

c-po commented:
@MergifyIo backport sagitta
